# Experimento numérico \#2 - Tema Aproximación SVD cambiando método de generación de matriz pseudo-aleatoria.

**Responsable:** León Garay 

## 0. Objetivo

Este documento pretende mostar los resultados obtenidos con la implementación realizada, en el contexto de la implementación de entradas normales con varianza cada vez más alta y matrices de $10^1$, $10^2$ y $10^3$.

**Nota:** El reporte general de resultados obtenidos con las implementaciones realizadas en este proyecto se puede consultar en el siguiente [vínculo](https://github.com/mno-2020-gh-classroom/ex-modulo-3-comp-matricial-svd-czammar/blob/master/results/Reporte_resultados.ipynb)

## 1. Consideraciones

Al respecto de los experimentos numéricos realizados para consolidar el presente reporte, tales se basan en las siguientes premisas:

[Pendiente: desarrollo]

En este sentido, en particular para cada experimento realizado, se reportan según resulte aplicable:

* los parámetros empleados en la simulaciones, 
* las dimensiones de las matrices y vectores involucrados, así como el prodecimiento pseudo-aleatorio que les dio origen, 
* 1) el tiempo involucrado en correr los experimentos, 
* 2) número de condición de las matrices pseudo-aleatorias, y 
* 3) el error relativo obtenido, para la solución de un sistema lineal de la forma $Ax=b$, es decir el cociente

$$|| Ax - b ||_2 / || b ||_2$$
* 4) La norma de $A$ y de la matriz obtenida al rearmar A tras multiplica las matrices resultantes de la aproximación de la descomposición SVD vía el algoritmo **One-Sided Jacobi**


### 1.1 Consideraciones sobre la infraestructura empleada

**Especificaciones de ambiente común de trabajo**

Para realizar el presente experimento numérico se ha empleado la imagen de docker basada en R del curso MNO 2020 (palmoreck/jupyterlab_r_kernel:1.1.0)

```
docker run --rm -v `pwd`:/datos --name jupyterlab_r_kernel_local -p 8888:8888 -d palmoreck/jupyterlab_r_kernel:1.1.0

```

*Nota:* el comando "-v \`pwd\`:/datos", permite montar el directorio actual en donde el usuario se encuentre situada en terminal como un volumen de la imagen de docker, dentro del directorio "/datos".

## 2. Experimento numérico

**Objetivo**

Modificar las entradas de la matriz y las dimensiones de los bloques para notar que efecto tiene sobre el tiempo de ejecución.

A tal respecto, se destaca que se realizó este experimentos buscando probar el comportamiento correcto de la función y medir los tiempo de ejecución.

**Cargamos codigo desarrollado previamente**

* **utils.R:** contiene las funciones auxiliares desarrolladas para el proyecto
* **00-load.R:** la implementación del método de eliminación por bloques, empleando la aproximación de la descomposición SVD vía el algoritmo **One-Sided Jacobi**.

In [1]:
## Instalamos paquetes
rm(list = ls())

paquetes <- c('matrixcalc')

instalar <- function(paquete) {
  if (!require(paquete,character.only = TRUE, quietly = TRUE, warn.conflicts = FALSE)) {
    install.packages(as.character(paquete), dependecies = TRUE, repos = "http://cran.us.r-project.org")
    library(paquete, character.only = TRUE, quietly = TRUE, warn.conflicts = FALSE)
  }
}


lapply(paquetes, instalar)

[[1]]
NULL

In [2]:
## Cargamos paquetes necesarios
library("matrixcalc")

#source("metadata.R")
source("utils.R")
source("00-load.R")

### 2.1 Experimento X.1  

Este experimento se basa en los siguientes parámetros 

| # |         Parámetros         | Valor/Rango de valores | Comentarios |
|:-:|:--------------------------:|:----------------------:|:-----------:|
| 1 |     Dimensiones de $A$     | $10x10$                | n=10        |
| 2 |        Tolerancia          | $10^{-8}$                |             |
| 3 |         Maxsweep           |   20                   |             |
| 4 |   Metodo pseudo-aleatorio  |  rnorm(n^2,0,1)        |             |

**Codigo:** Algoritmo que aplica el método de Jacobi para hacer la descomposición SVD de una matriz de entrada A (nxn).

In [3]:
set.seed(231)
n= 10**1

A = matrix(rnorm(n**2,0,1), ncol=n)
TOL = 10**-8
maxsweep <- 20
#Función aproximación SVD
svd<-svd_jacobi_aprox(A,TOL,maxsweep)

In [4]:
system.time(svd<-svd_jacobi_aprox(A,TOL,maxsweep))

   user  system elapsed 
  0.098   0.004   0.102 

In [5]:
#Regresamos a la matriz original usando SVD
A_prima = svd$U%*%svd$S%*%t(svd$V)

In [6]:
#Comprobamos que las matrices seán iguales.
identical(round(A,5),round(A_prima,5))

[1] TRUE

**Resultados**

Los resultados obtenidos se resumen a través de la siguiente tabla

| # |                Parámetros                | Valor/Rango de valores |            Comentarios            |
|:-:|:----------------------------------------:|:----------------------:|:---------------------------------:|
| 1 |            Tiempo de ejecución           |   0.098 s              |                                   |
| 2 |        $A = U*S*V^{T}$                   |   TRUE                 | La función sí regresa la matriz de entrada A

### 2.2  Experimento X.2 

Este experimento se basa en los siguientes parámetros 

| # |         Parámetros         | Valor/Rango de valores | Comentarios |
|:-:|:--------------------------:|:----------------------:|:-----------:|
| 1 |     Dimensiones de $A$     | $10x10$                | n=10^2        |
| 2 |        Tolerancia          | $10^{-8}$                |             |
| 3 |         Maxsweep           |   20                   |             |
| 4 |   Metodo pseudo-aleatorio  |  rnorm(n^2,0,10)        |             |

**Codigo:** Algoritmo que aplica el método de Jacobi para hacer la descomposición SVD de una matriz de entrada A (nxn).

In [7]:
set.seed(231)
n= 10**2

A = matrix(rnorm(n**2,0,10), ncol=n)
TOL = 10**-8
maxsweep <- 20
#Función aproximación SVD
svd<-svd_jacobi_aprox(A,TOL,maxsweep)

In [8]:
system.time(svd<-svd_jacobi_aprox(A,TOL,maxsweep))

   user  system elapsed 
 12.146   0.000  12.148 

In [9]:
#Regresamos a la matriz original usando SVD
A_prima = svd$U%*%svd$S%*%t(svd$V)

In [10]:
#Comprobamos que las matrices seán iguales.
identical(round(A,5),round(A_prima,5))

[1] TRUE

**Resultados**

Los resultados obtenidos se resumen a través de la siguiente tabla

| # |                Parámetros                | Valor/Rango de valores |            Comentarios            |
|:-:|:----------------------------------------:|:----------------------:|:---------------------------------:|
| 1 |            Tiempo de ejecución           |   12.146 s              |                                   |
| 2 |        $A = U*S*V^{T}$                   |   TRUE                 | La función sí regresa la matriz de entrada A

### 2.3  Experimento X.3 

Este experimento se basa en los siguientes parámetros 

| # |         Parámetros         | Valor/Rango de valores | Comentarios |
|:-:|:--------------------------:|:----------------------:|:-----------:|
| 1 |     Dimensiones de $A$     | $1000x1000$                | n=10^3        |
| 2 |        Tolerancia          | $10^{-8}$                |             |
| 3 |         Maxsweep           |   20                   |             |
| 4 |   Metodo pseudo-aleatorio  |  rnorm(n^2,0,100)        |             |

**Codigo:** Algoritmo que aplica el método de Jacobi para hacer la descomposición SVD de una matriz de entrada A (nxn).

In [13]:
set.seed(231)
n= 10**3

A = matrix(rnorm(n**2,0,100), ncol=n)
TOL = 10**-8
maxsweep <- 20
#Función aproximación SVD
svd<-svd_jacobi_aprox(A,TOL,maxsweep)

In [14]:
system.time(svd<-svd_jacobi_aprox(A,TOL,maxsweep))

    user   system  elapsed 
2330.452    0.940 2331.730 

In [15]:
#Regresamos a la matriz original usando SVD
A_prima = svd$U%*%svd$S%*%t(svd$V)

In [16]:
#Comprobamos que las matrices seán iguales.
identical(round(A,5),round(A_prima,5))

[1] TRUE

**Resultados**

Los resultados obtenidos se resumen a través de la siguiente tabla

| # |                Parámetros                | Valor/Rango de valores |            Comentarios            |
|:-:|:----------------------------------------:|:----------------------:|:---------------------------------:|
| 1 |            Tiempo de ejecución           |   2330.452 s              |                                   |
| 2 |        $A = U*S*V^{T}$                   |   TRUE                 | La función sí regresa la matriz de entrada A

**Resumen de resultados obtenidos en experimentos**

Los resultados obtenidos, se compararan a través de la siguiente [Pendiente: por favor ajustar tabla/grafica según experimento]:

| # | Experimento | Tiempo de desempeño | A = U * S * $V^T$ |
|:-:|:-----------:|:-------------------:|----------------|
| 1 |     X.1     |  0.098 s      |    True     |
| 2 |     X.2     | 12.146 s            |       True         |
| 3 |     X.3     | 2330.452 s           |         True       |

## 3. Principalez hallazgos

Al respecto, se destacan los siguientes hallazgos del experimento:

* Hallazgo 1: Los tiempos de ejecución aumentar de manera considerable al aumentar la dimención de la matriz.
* Hallazgo 2: El algoritmo fue consistente aumentando la dimención de la matriz y la varianza de las entradas, devolviendo la SVD correspondiente.
